### DISCUSSION TITLE
How to detect when a UNIX socket is closed when reading

### DISCUSSION THREAD

**Discussion by mikaelstaldal** [tokio-rs/tokio#6207] [Q&A]
I am developing a program which communicates with a Docker daemon, with HTTP over UNIX socket, using hyperlocal.

The problem I have is that it doesn't detect when the Docker daemon closes the socket. This is necessary when attaching to a container, since I want my program to exit when the container stops and it has consumed all output from the container (just like the `docker attach` command does). If I forcefully exit immediately when the container has stopped, I might miss some output.

I want this loop to exit when the socket is closed, but it doesn't:
https://github.com/mikaelstaldal/contained/blob/main/src/docker_client.rs#L197-L201

See whole project here: https://github.com/mikaelstaldal/contained

**Comment by Darksonn** [comment]
Unless you are stuck in `write_all` or `flush`, that code should work.

**Reply by mikaelstaldal** [reply]
Well, it doesn't. I just tried to comment out `write_all` and `flush`, just having an empty loop. It still doesn't exit.
